File: 604-2/A.cpp
#include <iostream>
#include <cstdio>
#include <string>
using namespace std;

char getnot(string lets) {
    if (lets.compare("a") == 0) {
        return 'b';
    } else if (lets.compare("b") == 0 || lets.compare("c") == 0) { 
        return 'a';
    } else if (lets.compare("ab") == 0 || lets.compare("ba") == 0 || lets.compare("aa") == 0 || lets.compare("bb") == 0) {
        return 'c';
    } else if (lets.compare("bc") == 0 || lets.compare("cb") == 0 || lets.compare("cc") == 0) {
        return 'a';
    } else if (lets.compare("ac") == 0 || lets.compare("ca") == 0) {
        return 'b';
    } else {
        return 'a';
    }
}

string generate(int length, string l, string r) {
    string ret = "";
    string prev = l;
    for (int i = 0; i < length; ++i) {
        if (i != length-1) {
            char ll = getnot(prev);
            ret += ll;
            prev = ll;
        } else {
            char ll = getnot(prev+r);
            ret += ll;
        }
    }
    
    return ret;
}

string find(string s) {
    string poss = "";
    int run = 0;
    for (auto let : s) {
        if (poss.compare("") == 0) {
            if (let == '?') {
                run += 1;
            } else {
                string r(1, let);
                poss += generate(run, "", r) + r;
                run = 0;
            }
        } else {
            if (let == '?') {
                run += 1;
            } else {
                string l(1, poss[poss.length()-1]);
                string r(1, let);
                poss += generate(run, l, r) + r;
                run = 0;
                if (poss[poss.length()-2] == poss[poss.length()-1]) {
                    return "-1";
                }
            }
        }
    }
    if (poss.compare("") != 0) {
        string l(1, poss[poss.length()-1]);
        poss += generate(run, l, "");
    } else {
        poss += generate(run, "", "");
    }
                    
    return poss;
}

int main() {
    int t;
    scanf("%d", &t);

    for (int i = 0; i < t; ++i) {
        string s;
        cin >> s;
        cout << find(s) << endl;
    }
}

File: 604-2/A.py
def getnot(lets):
    if lets == "a":
        return "b"
    elif lets == "b" or lets == "c":
        return "a"
    elif lets == "ab" or lets == "ba" or lets == "aa" or lets == "bb":
        return "c"
    elif lets == "bc" or lets == "cb" or lets == "cc":
        return "a"
    elif lets == "ac" or lets == "ca":
        return "b"
    else:
        return "a"

def generate(length, l, r):
    ret = ""
    prev = l
    for i in range(length):
        if i != length-1:
            ll = getnot(prev)
            ret += ll
            prev = ll
        else:
            ll = getnot(prev+r)
            ret += ll
    
    return ret

def find(s):
    poss = ""
    run = 0
    for let in s:
        if poss == "":
            if let == "?":
                run += 1
            else:
                poss += generate(run, "", let) + let
                run = 0
        else:
            if let == "?":
                run += 1
            else:
                poss += generate(run, poss[-1], let) + let
                run = 0
                if poss[-2] == poss[-1]:
                    return -1
    if poss != "":
        poss += generate(run, poss[-1], "")
    else:
        poss += generate(run, "", "")
                    
    return poss


t = int(input())
for i in range(t):
    s = input().rstrip()
    print(find(s))
